### PR TITLE
JBPM-9568 - Case Definition - Wrong location of the default WID file

### PR DESCRIPTION
**Pull request by hasys**
**JIRA**: [JBPM-9568 ](https://issues.redhat.com/browse/JBPM-9568)
**WAR**: https://drive.google.com/file/d/1fYIHG3xPppUBx7lLvCzVC9U8PaI6i8jT/view?usp=sharing

Hi @romartin, @adrielparedes, there is the fix for the issue. It also changes:
* WID creation moved to `global` directory instead of `resources` root, now when you create process you can add process in any package and all WIDs will be displayed
* added custom icons. I didn't add some service icons due to [JBPM-9568](https://issues.redhat.com/browse/JBPM-9568). I created small email icon but I didn't create small REST and others. CC @Josephblt
* WID synchronized with Stunner (primarily IO parameters)
* I removed code which by design supposed to create WID file in every new package. First of all it already doesn't work for initially created packages, and second there is no need to add same WID in every package, but User still able to add new WIDs in any package if necessary by `Create New Asset` -> `Work Item Definition`
* otherwise all should be the same as before and I expect no side effects.

Thank you!

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by hasys** [comment]
jenkins do fdb

**Review by adrielparedes — CHANGES_REQUESTED** [review]

**Comment by adrielparedes** [comment]
@hasys good job! Probably we will need to write some docs if the CaseManagementWID will be deprecated in order to use WorkItemDefinition.wid

**Comment by hasys** [comment]
Hi @adrielparedes,

thank you for prompt review. Slash `/` symbol in resources name instead of `separator` was used before in the resource name, not sure what is the difference, but I kept it in my changes so, not sure should we replace it by `separator`. See in removed code:
```java
private static final String CASE_WORK_DEFINITION_FILE = "/CaseWorkDefinitions.wid";
```



> @hasys good job! Probably we will need to write some docs if the CaseManagementWID will be deprecated in order to use WorkItemDefinition.wid

This change was internal for easier handling icons. Before `CaseWorkDefinitions` was a name of resource, but when user created a project Business Central created `WorkDefinition.wid` (see below removed code), so no changes from user point of view (instead of location of *.wid).

```java
private static final String WORK_DEFINITION_FILE = "WorkDefinition.wid";
private static final String CASE_WORK_DEFINITION_FILE = "/CaseWorkDefinitions.wid";

public void addWorkDefinitionsFileToProjectResources(KieModule kieModule, WorkspaceProject project, String separator) {
...
    Path resourcesPath = ioService.get(URI.create(resourcesPathStr + separator + WORK_DEFINITION_FILE));
    addWorkDefinitions(resourcesPath);
}

protected void addWorkDefinitions(Path location) {
    ...
    byte[] data = IOUtils.toByteArray(this.getClass().getResourceAsStream(CASE_WORK_DEFINITION_FILE));
    ioService.write(location, data);
    ...
```

So, I think to keep it as it was, no further changes in this PR required. WDYT?

**Comment by adrielparedes** [comment]
@hasys thanks for the explanation, in that case I don't see anything else to change. Should we include someone from QE?

**Comment by hasys** [comment]
Hi @LuboTerifaj, adding you as reviewer as discussed, thank you!

@adrielparedes good point, done. Thank you! :) Also, can I ask you to approve the PR if it is OK to you? Thank you!

**Review by adrielparedes — APPROVED** [review]

**Comment by adrielparedes** [comment]
@hasys approved! I forgot to do it before

File: jbpm-wb-integration/jbpm-wb-integration-backend/src/main/java/org/jbpm/workbench/wi/backend/server/casemgmt/service/CaseProjectServiceImpl.java [hunk 1]
```diff
@@ -140,48 +149,39 @@ public void configureNewCaseProject(WorkspaceProject project) {
...
-            Path resourcesPath = ioService.get(URI.create(resourcesPathStr + separator + WORK_DEFINITION_FILE));
-            addWorkDefinitions(resourcesPath);
-        }
+        logger.debug("Default WorkItemDefinitions added to {}",
+                     globalPath.toString());
     }
 
-    protected void addWorkDefinitions(Path location) {
+    protected void addWorkDefinitions(KieModule kieModule, String separator) {
+        String globalPath = Paths.convert(kieModule.getKModuleXMLPath())
```
Comment (romartin) [inline jbpm-wb-integration/jbpm-wb-integration-backend/src/main/java/org/jbpm/workbench/wi/backend/server/casemgmt/service/CaseProjectServiceImpl.java:162]: is there some option for obtaining the root path directly from the kieModule API?

Comment (hasys) [reply jbpm-wb-integration/jbpm-wb-integration-backend/src/main/java/org/jbpm/workbench/wi/backend/server/casemgmt/service/CaseProjectServiceImpl.java:162]: Hi @romartin I didn't found better way. If you will take a look above there are some ways to get root folder, but it actually returns root against the repo, so it is just `/` it is ok in some cases but for file creation I have to use the full path with schema location, repository name, space name and so on, so I used this approach. Let me know if you know how to achieve it in a better way. Thank you!

Comment (romartin) [reply jbpm-wb-integration/jbpm-wb-integration-backend/src/main/java/org/jbpm/workbench/wi/backend/server/casemgmt/service/CaseProjectServiceImpl.java:162]: ok no worries, I also don't know if there is a better way. That API anyway is a bit old and I think we shouldn't spend much time on figuring out this, let's go ahead with this code and let's see if it works properly anyway.
Thanks!

**Review by romartin — APPROVED** [review]

**Review by LuboTerifaj — CHANGES_REQUESTED** [review]
Hello @hasys ,

I have a few comments:

- Custom task "StartCaseInstance", respectively "Sub Case" has a different name in palette in comparison with node on canvas.

- Once you delete WorkDefinitions.wid file, it's not possible to manually create a new one located in Global folder.

- If the WorkDefinitions.wid file is deleted manually, after opening already exiting process/case or creating a new process/case, a new WID file is created but its content is different from the initial WorkDefinitions.wid file. If you delete also this WID file, a new WID file is never created.

- Business Rule task as well as Decision task are missing in Case Project by default.
	
- Email, Rest and WebService icons (png files) are missing.

- Just for the record, we should consider to change WID files placement in sample projects as well. It's not necessary to do it in this jira, but it should be reported in a follow-up one.

Thanks!

**Comment by hasys** [comment]
Hi @LuboTerifaj,

great exploratory job, thank you! Just want to mention that non of mentioned issues are part of [JBPM-9568](https://issues.redhat.com/browse/JBPM-9568) Acceptance Criteria or initial end to end use case. So feel free to report it as separate issues for further improvements and let's discuss do we have time to implement it now or not. Thank you!

@romartin, @adrielparedes, I think this PR is ready to be merged. Thank you guys, great job!

**Comment by romartin** [comment]
@LuboTerifaj can you please approve, so I'll merge? Thanks!

**Comment by LuboTerifaj** [comment]
Hey @hasys ,

even the issues I have found are not part of the acceptance criteria, they are not present on master. Some of them are even regressions.

Thanks!

**Review by LuboTerifaj — APPROVED** [review]
Hello @hasys ,

as discussed on chat, here are the reported issues:

- [JBPM-9604](https://issues.redhat.com/browse/JBPM-9604) Change WID files placement in sample projects
- [JBPM-9603](https://issues.redhat.com/browse/JBPM-9603) Different content in WID files
- [JBPM-9602](https://issues.redhat.com/browse/JBPM-9602) Rest and WebService tasks have only default icons
- [JBPM-9601](https://issues.redhat.com/browse/JBPM-9601) Business Rule and Decision tasks are missing in Case Project by default
- [JBPM-9600](https://issues.redhat.com/browse/JBPM-9600) Not possible to create a WID file in Global folder
- [JBPM-9599](https://issues.redhat.com/browse/JBPM-9599) StartCaseInstance node has a different name on canvas

Approving.

Thanks!

**Comment by hasys** [comment]
Great job, thank you @LuboTerifaj! Will put my notes as well.

@romartin, @adrielparedes, PR is ready for merge, thank you!

**Pull request merged**